ruby/neo4j/driver/internal/util/error_util.rb

```diff
@@ -0,0 +1,113 @@
+# frozen_string_literal: true
+
+module Neo4j::Driver::Internal::Util
+  class ErrorUtil
+    DEFAULT_CONN_TERMINATED_REASON = 'Please ensure that your database is listening on the correct host and port and '\
+      'that you have compatible encryption settings both on Neo4j server and driver. '\
+      'Note that the default encryption setting has changed in Neo4j 4.0.'.freeze
+
+    SEC_EXCEPTION_CODE_MAPPING = {
+      'neo.clienterror.security.unauthorized': Neo4j::Driver::Exceptions::AuthenticationException,
+      'neo.clienterror.security.authorizationexpired': Neo4j::Driver::Exceptions::AuthorizationExpiredException,
+      'neo.clienterror.security.tokenexpired': Neo4j::Driver::Exceptions::TokenExpiredException
+    }.freeze
+
+    class << self
+      def new_connection_terminated_error(reason)
+        reason ||= DEFAULT_CONN_TERMINATED_REASON
+        Neo4j::Driver::Exceptions::ServiceUnavailableException.new("Connection to the database terminated. #{reason}")
+      end
+
+      def new_result_consumed_error
+        Neo4j::Driver::Exceptions::ResultConsumedException.new(
+          'Cannot access records on this result any more as the result has already been consumed '\
+          'or the query runner where the result is created has already been closed.'
+        )
+      end
+
+      def new_neo4j_error(code, message)
+        exception_class = case extract_error_class(code)
+                          when 'ClientError'
+                            if extract_error_sub_class(code) == 'Security'
+                              SEC_EXCEPTION_CODE_MAPPING[code.downcase] || Neo4j::Driver::Exceptions::SecurityException
+                            else
+                              code.casecmp?('Neo.ClientError.Database.DatabaseNotFound') ? Neo4j::Driver::Exceptions::FatalDiscoveryException : Neo4j::Driver::Exceptions::ClientException
+                            end
+                          when 'TransientError'
+                            Neo4j::Driver::Exceptions::TransientException
+                          else
+                            Neo4j::Driver::Exceptions::DatabaseException
+                          end
+
+        exception_class.new(code, message)
+      end
+
+      def fatal?(error)
+        return true unless error.is_a?(Neo4j::Driver::Exceptions::Neo4jException)
+
+        error_code = error.code&.downcase
+        return true if protocol_violation_error?(error_code)
+        return false if client_or_transient_error?(error_code)
```

Comment:
@klobuczek what do you think of this comment?

should I change this code like below? It will be different than original but more accurate as per ruby standard.

```suggestion
        protocol_violation_error?(error_code)
```

Reply:
neither the implementation not the suggestion is correct

Reply:
now sure what you are referring to.

Reply:
both your 1st version and your suggested change do not behave like the java code, not even if you assume that nil is equivalent to false.